trojsten/rules/susi.py

```diff
@@ -0,0 +1,188 @@
+from .default import FinishedRoundsResultsRulesMixin as FinishedRounds
+
+SUSI_AGAT = "agat"
+SUSI_BLYSKAVICA = "blyskavica"
+SUSI_CIFERSKY_CECH = "cifersky cech"
+
+
+SUSI_AGAT_MAX_COEFFICIENT = 8
+SUSI_ELIGIBLE_FOR_TASK_BOUND = [8, 8, 1000, 1000, 1000, 1000, 1000]
+
+SUSI_CAMP_TYPE = "SuŠi sústredenie"
+
+SUSI_YEARS_OF_CAMPS_HISTORY = 10
+
+
+class SUSIResultsGenerator(CategoryTagKeyGeneratorMixin, ResultsGenerator):
+    def __init__(self, tag):
+        super(SUSIResultsGenerator, self).__init__(tag)
+        self.susi_camps = None
+        self.trojsten_camps = None
+        self.puzzlehunt_participations = None
+        self.has_graduated = None
+        self.coefficients = {}
+
+    def get_user_coefficient(self, user, round):
+        if user not in self.coefficients:
+            if not self.susi_camps:
+                self.prepare_coefficients(round)
+
+            successful_semesters = self.susi_camps.get(user.pk, 0)
+            puzzlehunt_participations = self.puzzlehunt_participations.get(user.pk, 0)
+            trojsten_camps = self.trojsten_camps.get(user.pk, 0)
+            self.coefficients[user] = (
+                5 * successful_semesters + 3 * puzzlehunt_participations + trojsten_camps
+            )
+
+        return self.coefficients[user]
+
+    def get_graduation_status(self, user, round):
+        if user not in self.has_graduated:
+            if not self.has_graduated:
+                self.prepare_coefficients(round)
+
+        return self.has_graduated[user]
+
+    def prepare_coefficients(self, round):
+        """
+        Fetch from the db number of successful semester and number of participation
+        in other Trojsten camps of each user and store them in dictionaries. The prepared
+        data in dictionaries are used to compute the coefficient of a given user.
+        We consider only events happened before given round, so the coefficients are computed
+        correct in older results.
+        """
+
+        self.trojsten_camps = dict(
+            EventParticipant.objects.filter(
+                Q(
+                    # TODO I think commenting next line will retrieve participations in
+                    # all trojsten camps
+                    # event__type__name=SUSI_CAMP_TYPE,
+                    event__end_time__lt=round.end_time,
+                    event__end_time__year__gte=round.end_time.year - SUSI_YEARS_OF_CAMPS_HISTORY,
+                ),
+                Q(going=True) | Q(type=EventParticipant.PARTICIPANT),
+            )
+            .values("user")
+            .annotate(camps=Count("event__semester", distinct=True))
+            .values_list("user", "camps")
+        )
+
+        # We ignore camps that happened before SUSI_YEARS_OF_CAMPS_HISTORY years, so we don't
+        # produce too big dictionaries of users.
+        self.susi_camps = dict(
+            EventParticipant.objects.filter(
+                Q(
+                    event__type__name=SUSI_CAMP_TYPE,
```

Comment:
aj ked EventParticipant asi nenarastie do takych rozmerov aby to malo pozorovatelny vykonovy dopad, ja by som radsej porovnaval podla `id` ako podla `name`